### ISSUE TITLE
Wrong refcount of lexicals defined in if(...)

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#7101]
Migrated from [rt.perl.org#26584](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=26584) (status was 'resolved')

Searchable as RT26584$

**Comment by p5pRT** [comment]
### From nospam-abuse@ilyaz.org

This is a bug report for perl from ilya\,
generated with the help of perlbug 1\.34 running under perl v5\.8\.2\.

Apparently\, a lexical defined as

&nbsp;  if \(my $obj = new \.\.\.\)

has 2 as its refcount \(instead of 1\)\.  Extra unit of refcount survives to the
end of the \`if' block\.

To reproduce&#8203;: the script embedded below produces

&nbsp;  before DESTROY1
&nbsp;  after DESTROY1
&nbsp;  before DESTROY2
&nbsp;  x\->DESTROY\(2\)
&nbsp;  after DESTROY2
&nbsp;  x\->DESTROY\(1\)
&nbsp;  after BLOCK

instead of destroying $obj1 near DESTROY1 markers \(similar to how $obj is
destroyed\)\.

Yours\,
Ilya

\#\!/usr/bin/perl \-w
use strict;

\{  package x;
&nbsp;   sub new \{ my $class = shift; bless \[shift\]\, $class \}
&nbsp;   sub DESTROY \{ my $self = shift; warn "x\->DESTROY\($self\->\[0\]\)\\n" \}
\}

if \(my $obj1 = x\->new\(1\)\) \{
&nbsp;  my $obj2 = x\->new\(2\);
&nbsp;  warn "before DESTROY1\\n";
&nbsp;  $obj1 = 0;
&nbsp;  warn "after DESTROY1\\n";
&nbsp;  warn "before DESTROY2\\n";
&nbsp;  $obj2 = 0;
&nbsp;  warn "after DESTROY2\\n";
\}
warn "after BLOCK\\n";
\_\_END\_\_



---
Flags&#8203;:
&nbsp;    category=core
&nbsp;    severity=medium


---
Site configuration information for perl v5\.8\.2&#8203;:

Configured by vera at Sat Nov  8 12&#8203;:23&#8203;:19 PST 2003\.

Summary of my perl5 \(revision 5\.0 version 8 subversion 2\) configuration&#8203;:
&nbsp;  Platform&#8203;:
&nbsp;    osname=os2\, osvers=2\.30\, archname=os2
&nbsp;    uname='os2 ia\-ia 2 2\.30 i386 '
&nbsp;    config\_args='\-des \-D prefix=i&#8203;:/perllib'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
&nbsp;    useperlio=define d\_sfio=undef uselargefiles=define usesocks=undef
&nbsp;    use64bitint=undef use64bitall=undef uselongdouble=undef
&nbsp;    usemymalloc=y\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='gcc'\, ccflags ='\-Zomf \-Zmt \-DDOSISH \-DOS2=2 \-DEMBED \-I\. \-D\_EMX\_CRT\_REV\_=60'\,
&nbsp;    optimize='\-O2 \-fomit\-frame\-pointer \-malign\-loops=2 \-malign\-jumps=2 \-malign\-functions=2 \-s'\,
&nbsp;    cppflags='\-Zomf \-Zmt \-DDOSISH \-DOS2=2 \-DEMBED \-I\. \-D\_EMX\_CRT\_REV\_=60'
&nbsp;    ccversion=''\, gccversion='2\.8\.1'\, gccosandvers=''
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=1234
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=12
&nbsp;    ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=4
&nbsp;    alignbytes=4\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='gcc'\, ldflags ='\-Zexe \-Zomf \-Zmt \-Zcrtdll \-Zstack 32000 \-Zlinker /e&#8203;:2'
&nbsp;    libpth=i&#8203;:/emx\.add/lib i&#8203;:/emx/lib i&#8203;:/emx\.f77/lib D&#8203;:/DEVTOOLS/OPENGL/LIB I&#8203;:/JAVA11/LIB i&#8203;:/emx/lib/mt
&nbsp;    libs=\-lsocket \-lm \-lbsd \-lcrypt
&nbsp;    perllibs=\-lsocket \-lm \-lbsd \-lcrypt
&nbsp;    libc=i&#8203;:/emx/lib/mt/c\_import\.lib\, so=dll\, useshrplib=true\, libperl=libperl\.lib
&nbsp;    gnulibc\_version=''
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=dll\, d\_dlsymun=undef\, ccdlflags=' '
&nbsp;    cccdlflags='\-Zdll'\, lddlflags='\-Zdll \-Zomf \-Zmt \-Zcrtdll \-Zlinker /e&#8203;:2'

Locally applied patches&#8203;:
&nbsp;    



---
@&#8203;INC for perl v5\.8\.2&#8203;:
&nbsp;    i&#8203;:/perllib/lib/5\.8\.2/os2
&nbsp;    i&#8203;:/perllib/lib/5\.8\.2
&nbsp;    i&#8203;:/perllib/lib/site\_perl/5\.8\.2/os2
&nbsp;    i&#8203;:/perllib/lib/site\_perl/5\.8\.2
&nbsp;    i&#8203;:/perllib/lib/site\_perl
&nbsp;    \.



---
Environment for perl v5\.8\.2&#8203;:
&nbsp;    HOME=j&#8203;:/home
&nbsp;    LANG=EN\_US
&nbsp;    LANGUAGE \(unset\)
&nbsp;    LD\_LIBRARY\_PATH \(unset\)
&nbsp;    LOGDIR \(unset\)
&nbsp;    PATH=i&#8203;:\\NETSCAPE\\PROGRAM;d&#8203;:\\OPENDOC\\EPM;d&#8203;:\\OPENDOC\\OREXX;d&#8203;:\\OPENDOC\\BIN;i&#8203;:\\DRV;i&#8203;:\\TOOLKIT\\BIN\\OSARED;i&#8203;:\\TOOLKIT\\SOM\\COMMON;i&#8203;:\\TOOLKIT\\SOM\\BIN;i&#8203;:\\TOOLKIT\\BIN;D&#8203;:\\OS2;D&#8203;:\\OS2\\SYSTEM;i&#8203;:\\VABASIC\\SOM\\BIN;D&#8203;:\\OS2\\INSTALL;D&#8203;:\\;D&#8203;:\\OS2\\MDOS;D&#8203;:\\OS2\\APPS;i&#8203;:\\BIN;i&#8203;:\\UTILS;i&#8203;:\\EMACS\\19\.33\\BIN;i&#8203;:\\emx\.add\\BIN;i&#8203;:\\EMX\\BIN;i&#8203;:\\emx\.f77\\BIN;i&#8203;:\\EMTEX\\bin;i&#8203;:\\ckermit;I&#8203;:\\PAGETURN;d&#8203;:\\tcpip\\bin;D&#8203;:\\TCPIP\\UMAIL;D&#8203;:\\VIEWER\\BIN;I&#8203;:\\XFree86\\bin;i&#8203;:\\VABASIC\\BIN;i&#8203;:\\SMART;I&#8203;:\\Java11\\rmi\-iiop\\bin;I&#8203;:\\JAVA11\\BIN;d&#8203;:\\JAVAOS2\\bin;D&#8203;:\\MMOS2;I&#8203;:\\JAVA11\\ICATJAVA\\BIN;I&#8203;:\\gs\\tools;I&#8203;:\\gs\\gs7\.00\\bin;i&#8203;:\\IBMWORKS;I&#8203;:\\Office51;;I&#8203;:\\zax\\bin\\;I&#8203;:\\zax;I&#8203;:\\zax\\nc;I&#8203;:\\zax\\NARC;f&#8203;:\\nu;c&#8203;:\\dos;
&nbsp;    PERLLIB\_PREFIX=f&#8203;:/perllib;i&#8203;:/perllib
&nbsp;    PERL\_BADLANG \(unset\)
&nbsp;    PERL\_SH\_DIR=i&#8203;:/bin
&nbsp;    SHELL \(unset\)

**Comment by p5pRT** [comment]
### From @iabyn

On Thu\, Feb 12\, 2004 at 06&#8203;:50&#8203;:24AM \-0000\, Ilya Zakharevich wrote&#8203;:
> Apparently\, a lexical defined as
> 
>   if \(my $obj = new \.\.\.\)
> 
> has 2 as its refcount \(instead of 1\)\.  Extra unit of refcount survives to the
> end of the \`if' block\.
> 
> To reproduce&#8203;: the script embedded below produces
> 
>   before DESTROY1
>   after DESTROY1
>   before DESTROY2
>   x\->DESTROY\(2\)
>   after DESTROY2
>   x\->DESTROY\(1\)
>   after BLOCK
> 
> instead of destroying $obj1 near DESTROY1 markers \(similar to how $obj is
> destroyed\)\.
> 
> Yours\,
> Ilya
> 
> \#\!/usr/bin/perl \-w
> use strict;
> 
> \{  package x;
>    sub new \{ my $class = shift; bless \[shift\]\, $class \}
>    sub DESTROY \{ my $self = shift; warn "x\->DESTROY\($self\->\[0\]\)\\n" \}
> \}
> 
> if \(my $obj1 = x\->new\(1\)\) \{
>   my $obj2 = x\->new\(2\);
>   warn "before DESTROY1\\n";
>   $obj1 = 0;
>   warn "after DESTROY1\\n";
>   warn "before DESTROY2\\n";
>   $obj2 = 0;
>   warn "after DESTROY2\\n";
> \}
> warn "after BLOCK\\n";

Its actually nothing to do with lexicals; its just that temporary values
aren't freed until the end of the statement they appear in; in this case
that's at the end of the if block\, eg

&nbsp;    use strict;

&nbsp;    \{  package x;
&nbsp;       sub new \{ my $class = shift; bless \[shift\]\, $class \}
&nbsp;       sub DESTROY \{ my $self = shift; warn "x\->DESTROY\($self\->\[0\]\)\\n" \}
&nbsp;    \}

&nbsp;    if \(x\->new\(1\)\) \{
&nbsp;    warn "in if block\\n";
&nbsp;    \}   
&nbsp;    warn "after BLOCK\\n"; 

outputs&#8203;:

&nbsp;    in if block
&nbsp;    x\->DESTROY\(1\)
&nbsp;    after BLOCK



\-\- 
Wesley Crusher gets beaten up by his classmates for being a smarmy git\,
and consequently has a go at making some friends of his own age for a
change\.
&nbsp;    \-\- Things That Never Happen in "Star Trek" \#18

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From nick.ing-simmons@elixent.com

Ilya Zakharevich \<perl5\-porters@&#8203;perl\.org> writes&#8203;:
>\# New Ticket Created by  Ilya Zakharevich 
>\# Please include the string&#8203;:  \[perl \#26584\]
>\# in the subject line of all future correspondence about this issue\. 
>\# \<URL&#8203;: http&#8203;://rt\.perl\.org&#8203;:80/rt3/Ticket/Display\.html?id=26584 >
>
>
>This is a bug report for perl from ilya\,
>generated with the help of perlbug 1\.34 running under perl v5\.8\.2\.
>
>Apparently\, a lexical defined as
>
>  if \(my $obj = new \.\.\.\)
>
>has 2 as its refcount \(instead of 1\)\.  Extra unit of refcount survives to the
>end of the \`if' block\.

I think this was deliberate\.

>
>To reproduce&#8203;: the script embedded below produces
>
>  before DESTROY1
>  after DESTROY1
>  before DESTROY2
>  x\->DESTROY\(2\)
>  after DESTROY2
>  x\->DESTROY\(1\)
>  after BLOCK
>
>instead of destroying $obj1 near DESTROY1 markers \(similar to how $obj is
>destroyed\)\.
>
>Yours\,
>Ilya
>
>\#\!/usr/bin/perl \-w
>use strict;
>
>\{  package x;
>   sub new \{ my $class = shift; bless \[shift\]\, $class \}
>   sub DESTROY \{ my $self = shift; warn "x\->DESTROY\($self\->\[0\]\)\\n" \}
>\}
>
>if \(my $obj1 = x\->new\(1\)\) \{
>  my $obj2 = x\->new\(2\);
>  warn "before DESTROY1\\n";
>  $obj1 = 0;
>  warn "after DESTROY1\\n";
>  warn "before DESTROY2\\n";
>  $obj2 = 0;
>  warn "after DESTROY2\\n";
>\}
>warn "after BLOCK\\n";
>\_\_END\_\_
>
>\-\-\-
>Flags&#8203;:
>    category=core
>    severity=medium
>\-\-\-
>Site configuration information for perl v5\.8\.2&#8203;:
>
>Configured by vera at Sat Nov  8 12&#8203;:23&#8203;:19 PST 2003\.
>
>Summary of my perl5 \(revision 5\.0 version 8 subversion 2\) configuration&#8203;:
>  Platform&#8203;:
>    osname=os2\, osvers=2\.30\, archname=os2
>    uname='os2 ia\-ia 2 2\.30 i386 '
>    config\_args='\-des \-D prefix=i&#8203;:/perllib'
>    hint=recommended\, useposix=true\, d\_sigaction=define
>    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
>    useperlio=define d\_sfio=undef uselargefiles=define usesocks=undef
>    use64bitint=undef use64bitall=undef uselongdouble=undef
>    usemymalloc=y\, bincompat5005=undef
>  Compiler&#8203;:
>    cc='gcc'\, ccflags ='\-Zomf \-Zmt \-DDOSISH \-DOS2=2 \-DEMBED \-I\. \-D\_EMX\_CRT\_REV\_=60'\,
>    optimize='\-O2 \-fomit\-frame\-pointer \-malign\-loops=2 \-malign\-jumps=2 \-malign\-functions=2 \-s'\,
>    cppflags='\-Zomf \-Zmt \-DDOSISH \-DOS2=2 \-DEMBED \-I\. \-D\_EMX\_CRT\_REV\_=60'
>    ccversion=''\, gccversion='2\.8\.1'\, gccosandvers=''
>    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=1234
>    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=12
>    ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=4
>    alignbytes=4\, prototype=define
>  Linker and Libraries&#8203;:
>    ld='gcc'\, ldflags ='\-Zexe \-Zomf \-Zmt \-Zcrtdll \-Zstack 32000 \-Zlinker /e&#8203;:2'
>    libpth=i&#8203;:/emx\.add/lib i&#8203;:/emx/lib i&#8203;:/emx\.f77/lib D&#8203;:/DEVTOOLS/OPENGL/LIB I&#8203;:/JAVA11/LIB i&#8203;:/emx/lib/mt
>    libs=\-lsocket \-lm \-lbsd \-lcrypt
>    perllibs=\-lsocket \-lm \-lbsd \-lcrypt
>    libc=i&#8203;:/emx/lib/mt/c\_import\.lib\, so=dll\, useshrplib=true\, libperl=libperl\.lib
>    gnulibc\_version=''
>  Dynamic Linking&#8203;:
>    dlsrc=dl\_dlopen\.xs\, dlext=dll\, d\_dlsymun=undef\, ccdlflags=' '
>    cccdlflags='\-Zdll'\, lddlflags='\-Zdll \-Zomf \-Zmt \-Zcrtdll \-Zlinker /e&#8203;:2'
>
>Locally applied patches&#8203;:
>    
>
>\-\-\-
>@&#8203;INC for perl v5\.8\.2&#8203;:
>    i&#8203;:/perllib/lib/5\.8\.2/os2
>    i&#8203;:/perllib/lib/5\.8\.2
>    i&#8203;:/perllib/lib/site\_perl/5\.8\.2/os2
>    i&#8203;:/perllib/lib/site\_perl/5\.8\.2
>    i&#8203;:/perllib/lib/site\_perl
>    \.
>
>\-\-\-
>Environment for perl v5\.8\.2&#8203;:
>    HOME=j&#8203;:/home
>    LANG=EN\_US
>    LANGUAGE \(unset\)
>    LD\_LIBRARY\_PATH \(unset\)
>    LOGDIR \(unset\)
>    PATH=i&#8203;:\\NETSCAPE\\PROGRAM;d&#8203;:\\OPENDOC\\EPM;d&#8203;:\\OPENDOC\\OREXX;d&#8203;:\\OPENDOC\\BIN;i&#8203;:\\DRV;i&#8203;:\\TOOLKIT\\BIN\\OSARED;i&#8203;:\\TOOLKIT\\SOM\\COMMON;i&#8203;:\\TOOLKIT\\SOM\\BIN;i&#8203;:\\TOOLKIT\\BIN;D&#8203;:\\OS2;D&#8203;:\\OS2\\SYSTEM;i&#8203;:\\VABASIC\\SOM\\BIN;D&#8203;:\\OS2\\INSTALL;D&#8203;:\\;D&#8203;:\\OS2\\MDOS;D&#8203;:\\OS2\\APPS;i&#8203;:\\BIN;i&#8203;:\\UTILS;i&#8203;:\\EMACS\\19\.33\\BIN;i&#8203;:\\emx\.add\\BIN;i&#8203;:\\EMX\\BIN;i&#8203;:\\emx\.f77\\BIN;i&#8203;:\\EMTEX\\bin;i&#8203;:\\ckermit;I&#8203;:\\PAGETURN;d&#8203;:\\tcpip\\bin;D&#8203;:\\TCPIP\\UMAIL;D&#8203;:\\VIEWER\\BIN;I&#8203;:\\XFree86\\bin;i&#8203;:\\VABASIC\\BIN;i&#8203;:\\SMART;I&#8203;:\\Java11\\rmi\-iiop\\bin;I&#8203;:\\JAVA11\\BIN;d&#8203;:\\JAVAOS2\\bin;D&#8203;:\\MMOS2;I&#8203;:\\JAVA11\\ICATJAVA\\BIN;I&#8203;:\\gs\\tools;I&#8203;:\\gs\\gs7\.00\\bin;i&#8203;:\\IBMWORKS;I&#8203;:\\Office51;;I&#8203;:\\zax\\bin\\;I&#8203;:\\zax;I&#8203;:\\zax\\nc;I&#8203;:\\zax\\NARC;f&#8203;:\\nu;c&#8203;:\\dos;
>    PERLLIB\_PREFIX=f&#8203;:/perllib;i&#8203;:/perllib
>    PERL\_BADLANG \(unset\)
>    PERL\_SH\_DIR=i&#8203;:/bin
>    SHELL \(unset\)

**Comment by p5pRT** [comment]
@iabyn - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From nospam-abuse@ilyaz.org

On Sat\, Feb 14\, 2004 at 02&#8203;:48&#8203;:38AM \-0000\, Dave Mitchell via RT wrote&#8203;:
> > Apparently\, a lexical defined as
> > 
> >   if \(my $obj = new \.\.\.\)
> > 
> > has 2 as its refcount \(instead of 1\)\.  Extra unit of refcount survives to the
> > end of the \`if' block\.

> Its actually nothing to do with lexicals; its just that temporary values
> aren't freed until the end of the statement they appear in; in this case
> that's at the end of the if block\, eg
> 
>     use strict;
> 
>     \{  package x;
>        sub new \{ my $class = shift; bless \[shift\]\, $class \}
>        sub DESTROY \{ my $self = shift; warn "x\->DESTROY\($self\->\[0\]\)\\n" \}
>     \}
> 
>     if \(x\->new\(1\)\) \{
>     warn "in if block\\n";
>     \}   
>     warn "after BLOCK\\n"; 

This is probably a correct analysis; however\, it just shows that the
scope of the bug is wider than originally stated\.  Do not you agree?

Yours\,
Ilya

**Comment by p5pRT** [comment]
### From @ysth

On Tue\, Mar 30\, 2004 at 01&#8203;:03&#8203;:00PM \-0800\, Ilya Zakharevich \<nospam\-abuse@&#8203;ilyaz\.org> wrote&#8203;:
> On Sat\, Feb 14\, 2004 at 02&#8203;:48&#8203;:38AM \-0000\, Dave Mitchell via RT wrote&#8203;:
> > > Apparently\, a lexical defined as
> > > 
> > >   if \(my $obj = new \.\.\.\)
> > > 
> > > has 2 as its refcount \(instead of 1\)\.  Extra unit of refcount survives to the
> > > end of the \`if' block\.
> 
> > Its actually nothing to do with lexicals; its just that temporary values
> > aren't freed until the end of the statement they appear in; in this case
> > that's at the end of the if block\, eg
> > 
> >     use strict;
> > 
> >     \{  package x;
> >        sub new \{ my $class = shift; bless \[shift\]\, $class \}
> >        sub DESTROY \{ my $self = shift; warn "x\->DESTROY\($self\->\[0\]\)\\n" \}
> >     \}
> > 
> >     if \(x\->new\(1\)\) \{
> >     warn "in if block\\n";
> >     \}   
> >     warn "after BLOCK\\n"; 
> 
> This is probably a correct analysis; however\, it just shows that the
> scope of the bug is wider than originally stated\.  Do not you agree?

Reminds me of this&#8203;:
&nbsp;  $ perl \-we'sub TIESCALAR \{ bless \{\} \} tie $x\, "main"; untie $x if tied $x'
&nbsp;  untie attempted while 1 inner references still exist at \-e line 1\.

Ilya\, do you have a mechanism in mind for freeing temps any earlier than
the end of the statement?

**Comment by p5pRT** [comment]
### From nospam-abuse@ilyaz.org

On Tue\, Mar 30\, 2004 at 09&#8203;:38&#8203;:13PM \-0000\, Yitzchak Scott\-Thoennes via RT wrote&#8203;:
> > > > Apparently\, a lexical defined as
> > > > 
> > > >   if \(my $obj = new \.\.\.\)
> > > > 
> > > > has 2 as its refcount \(instead of 1\)\.  Extra unit of refcount survives to the
> > > > end of the \`if' block\.
> > 
> > > Its actually nothing to do with lexicals; its just that temporary values
> > > aren't freed until the end of the statement they appear in; in this case

> > This is probably a correct analysis; however\, it just shows that the
> > scope of the bug is wider than originally stated\.  Do not you agree?

> Reminds me of this&#8203;:
>   $ perl \-we'sub TIESCALAR \{ bless \{\} \} tie $x\, "main"; untie $x if tied $x'
>   untie attempted while 1 inner references still exist at \-e line 1\.

This one may be harder to tackle\.\.\.

> Ilya\, do you have a mechanism in mind for freeing temps any earlier than
> the end of the statement?

I do not think so; however\, I do not see why the end of AAA in

&nbsp;  if \(AAA\) \{
&nbsp;    BBB
&nbsp;  \}

should be \*after\* BBB \(unless BBB is judged to be simple enough so
that a block is optimized away\)\.

Yours\,
Ilya

**Comment by p5pRT** [comment]
### From @lizmat

At 13&#8203;:03 \-0800 3/30/04\, Ilya Zakharevich wrote&#8203;:
>This is probably a correct analysis; however\, it just shows that the
>scope of the bug is wider than originally stated\.  Do not you agree?

My version of the problem&#8203;:

foreach \(1\.\.2\) \{
&nbsp;     print "Before $\_\\n";
&nbsp;     if \(my $object = Foo\->new\) \{
&nbsp;     \}
&nbsp;     print "After $\_\\n";
\}
print "After foreach\\n";

package Foo;
sub new \{ my $self = bless \{\}\,shift; print "CREATED $self\\n"; $self \}
sub DESTROY \{ print "DESTROYED $\_\[0\]\\n" \}

outputs&#8203;:

Before 1
CREATED Foo=HASH\(0x1066e4\)
After 1
DESTROYED Foo=HASH\(0x1066e4\)
Before 2
CREATED Foo=HASH\(0x1066e4\)
After 2
DESTROYED Foo=HASH\(0x1066e4\)
After foreach


But I would expect&#8203;:

Before 1
CREATED Foo=HASH\(0x1066e4\)
DESTROYED Foo=HASH\(0x1066e4\)
After 1
Before 2
CREATED Foo=HASH\(0x1066e4\)
DESTROYED Foo=HASH\(0x1066e4\)
After 2
After foreach


Although the destruction of the object is later than expected\, it 
still is destroyed for each loop iteration \(which would be a severe 
problem if that wouldn't happen\, as it would leak for a very common 
idiom\)\.


What I think is the worst case\, though\, is when the if \(my \.\.\.\) 
construct is used at the package level\.  In that case the memory 
won't be released / object destroyed until global destruction\!  This 
could be quite a surprise\!  For example \(using the same Foo package\)

print "Before\\n";
if \(my $object = Foo\->new\) \{
\}
print "After\\n";

outputs&#8203;:

Before
CREATED Foo=HASH\(0x103638\)
After
DESTROYED Foo=HASH\(0x103638\)

rather than what I expect&#8203;:

Before
CREATED Foo=HASH\(0x103638\)
DESTROYED Foo=HASH\(0x103638\)
After


Liz

**Comment by p5pRT** [comment]
### From @iabyn

On Tue\, Mar 30\, 2004 at 01&#8203;:49&#8203;:01PM \-0800\, Ilya Zakharevich wrote&#8203;:
> I do not think so; however\, I do not see why the end of AAA in
> 
>   if \(AAA\) \{
>     BBB
>   \}
> 
> should be \*after\* BBB \(unless BBB is judged to be simple enough so
> that a block is optimized away\)\.

&nbsp;    if \(AAA\) \{ BBB \}
&nbsp;    
is equivalent to

&nbsp;    AAA && do \{ BBB \}

Temps created by AAA get released at the end of the statement; temps
created in BBB get released at the end of BBB's block; I cant see anything
wrong with that\.

\-\- 
Monto Blanco\.\.\. scorchio\!

**Comment by p5pRT** [comment]
### From @iabyn

On Tue\, Mar 30\, 2004 at 11&#8203;:58&#8203;:10PM \+0200\, Elizabeth Mattijsen wrote&#8203;:
> My version of the problem&#8203;:
> 
> foreach \(1\.\.2\) \{
>     print "Before $\_\\n";
>     if \(my $object = Foo\->new\) \{
>     \}
>     print "After $\_\\n";
> \}
> print "After foreach\\n";
> 
> package Foo;
> sub new \{ my $self = bless \{\}\,shift; print "CREATED $self\\n"; $self \}
> sub DESTROY \{ print "DESTROYED $\_\[0\]\\n" \}
> 
> outputs&#8203;:
> 
> Before 1
> CREATED Foo=HASH\(0x1066e4\)
> After 1
> DESTROYED Foo=HASH\(0x1066e4\)
> Before 2
> CREATED Foo=HASH\(0x1066e4\)
> After 2
> DESTROYED Foo=HASH\(0x1066e4\)
> After foreach


This is because the \(\) of the if doesn't introduce a new run\-time
scope\. I'm sure I mentioned this recently in reply to another bug report\.

\-\- 
"The GPL violates the U\.S\. Constitution\, together with copyright\,
antitrust and export control laws"
&nbsp;    \-\- SCO smoking crack again\.

**Comment by p5pRT** [comment]
### From nospam-abuse@ilyaz.org

On Wed\, Mar 31\, 2004 at 12&#8203;:00&#8203;:51AM \+0100\, Dave Mitchell wrote&#8203;:
> On Tue\, Mar 30\, 2004 at 01&#8203;:49&#8203;:01PM \-0800\, Ilya Zakharevich wrote&#8203;:
> > I do not think so; however\, I do not see why the end of AAA in
> > 
> >   if \(AAA\) \{
> >     BBB
> >   \}
> > 
> > should be \*after\* BBB \(unless BBB is judged to be simple enough so
> > that a block is optimized away\)\.
> 
>     if \(AAA\) \{ BBB \}
>     
> is equivalent to
> 
>     AAA && do \{ BBB \}

But it should not be\.  It should be at least

&nbsp;  do \{AAA\} and do \{BBB\}

but probably yet something else which conforms to people's expectations\.

> Temps created by AAA get released at the end of the statement; temps
> created in BBB get released at the end of BBB's block; I cant see anything
> wrong with that\.

The fact that lifetime is extended is wrong\.

\[The most pleasant feature of logic is that there are so many to choose from\.\]

Hope this helps\,
Ilya

**Comment by p5pRT** [comment]
### From @rgs

Quoting Ilya Zakharevich \<nospam\-abuse@&#8203;ilyaz\.org>&#8203;:
> On Wed\, Mar 31\, 2004 at 12&#8203;:00&#8203;:51AM \+0100\, Dave Mitchell wrote&#8203;:
> >     if \(AAA\) \{ BBB \}
> >     
> > is equivalent to
> > 
> >     AAA && do \{ BBB \}
> 
> But it should not be\.  It should be at least
> 
>   do \{AAA\} and do \{BBB\}
> 
> but probably yet something else which conforms to people's expectations\.

You probably meant
&nbsp;    do \{ AAA and do \{ BBB \} \}
didn't you?
Besides this\, I don't really like the idea of introducing new
hidden scopes just for this kind of problem\. They have a runtime
cost\.

**Comment by p5pRT** [comment]
### From @ysth

On Wed\, Mar 31\, 2004 at 10&#8203;:04&#8203;:05AM \+0200\, Rafael Garcia\-Suarez \<rgarciasuarez@&#8203;free\.fr> wrote&#8203;:
> Quoting Ilya Zakharevich \<nospam\-abuse@&#8203;ilyaz\.org>&#8203;:
> > On Wed\, Mar 31\, 2004 at 12&#8203;:00&#8203;:51AM \+0100\, Dave Mitchell wrote&#8203;:
> > >     if \(AAA\) \{ BBB \}
> > >     
> > > is equivalent to
> > > 
> > >     AAA && do \{ BBB \}
> > 
> > But it should not be\.  It should be at least
> > 
> >   do \{AAA\} and do \{BBB\}
> > 
> > but probably yet something else which conforms to people's expectations\.
> 
> You probably meant
>     do \{ AAA and do \{ BBB \} \}
> didn't you?

I don't think that's what he meant\.  That would solve the "problem" that
Liz showed with scope of lexicals \(which I think works as designed\)\, but
not IIUC the different "problem" Ilya is reporting\.

do \{AAA\} and do \{BBB\} was presumably suggested to free temps from AAA
before executing BBB\, but doesn't actually work for that\.

> Besides this\, I don't really like the idea of introducing new
> hidden scopes just for this kind of problem\. They have a runtime
> cost\.

If it is possible to do something\, it needs to be done within pp\_and
itself\, and only when \*not\* short\-circuiting\.

**Comment by p5pRT** [comment]
### From nospam-abuse@ilyaz.org

On Wed\, Mar 31\, 2004 at 08&#8203;:05&#8203;:40AM \-0000\, Rafael Garcia\-Suarez via RT wrote&#8203;:
> > But it should not be\.  It should be at least
> > 
> >   do \{AAA\} and do \{BBB\}
> > 
> > but probably yet something else which conforms to people's expectations\.
> 
> You probably meant
>     do \{ AAA and do \{ BBB \} \}
> didn't you?

No\.

> Besides this\, I don't really like the idea of introducing new
> hidden scopes just for this kind of problem\. They have a runtime
> cost\.

Even if this were correct\, the cost of having too long a lifetime is
much worse&#8203;: a lost programmer time\.  Anyway\, what we need is not an
extra scope\, but an extra free\-tmps point\, like in the beginning of a
statement\.

Do you want to replace most of ';' in your programs to '\,' to increase
speed?  This is the same optimization\.\.\.

Hope this helps\,
Ilya

**Comment by p5pRT** [comment]
### From nospam-abuse@ilyaz.org

On Wed\, Mar 31\, 2004 at 11&#8203;:25&#8203;:38AM \-0800\, Yitzchak Scott\-Thoennes wrote&#8203;:
> do \{AAA\} and do \{BBB\} was presumably suggested to free temps from AAA
> before executing BBB\, but doesn't actually work for that\.

Right\.  It should have been something like&#8203;:

&nbsp;  do \{my $res = \!\!AAA; $res\} and do \{BBB\}

\(sigh; in C it is much easier than in Perl\.\.\.\)\.

Thanks\,
Ilya

**Comment by p5pRT** [comment]
### From @ysth

On Wed\, Mar 31\, 2004 at 04&#8203;:51&#8203;:44PM \-0800\, Ilya Zakharevich \<nospam\-abuse@&#8203;ilyaz\.org> wrote&#8203;:
> On Wed\, Mar 31\, 2004 at 11&#8203;:25&#8203;:38AM \-0800\, Yitzchak Scott\-Thoennes wrote&#8203;:
> > do \{AAA\} and do \{BBB\} was presumably suggested to free temps from AAA
> > before executing BBB\, but doesn't actually work for that\.
> 
> Right\.  It should have been something like&#8203;:
> 
>   do \{my $res = \!\!AAA; $res\} and do \{BBB\}
> 
> \(sigh; in C it is much easier than in Perl\.\.\.\)\.

That only works if the overall statement is in void context or AAA is true\.
Otherwise\, AAA is the resulting value of the statement\.

**Comment by p5pRT** [comment]
### From nospam-abuse@ilyaz.org

On Wed\, Mar 31\, 2004 at 08&#8203;:18&#8203;:21PM \-0800\, Yitzchak Scott\-Thoennes wrote&#8203;:
> On Wed\, Mar 31\, 2004 at 04&#8203;:51&#8203;:44PM \-0800\, Ilya Zakharevich \<nospam\-abuse@&#8203;ilyaz\.org> wrote&#8203;:
> > On Wed\, Mar 31\, 2004 at 11&#8203;:25&#8203;:38AM \-0800\, Yitzchak Scott\-Thoennes wrote&#8203;:
> > > do \{AAA\} and do \{BBB\} was presumably suggested to free temps from AAA
> > > before executing BBB\, but doesn't actually work for that\.
> > 
> > Right\.  It should have been something like&#8203;:
> > 
> >   do \{my $res = \!\!AAA; $res\} and do \{BBB\}
> > 
> > \(sigh; in C it is much easier than in Perl\.\.\.\)\.

> That only works if the overall statement is in void context or AAA is true\.
> Otherwise\, AAA is the resulting value of the statement\.

Do you still consider

&nbsp;  if \(AAA\) \{BBB\}

a statement?  It is not\.  AAA is a separate statement; BBB consists of
separate statements\.

But anyway\, your analysis finally pointed out \*some\* reason why the
current behaviour makes sense in \*some\* situations\.  Obviously\, I
missed this; thanks\.

Since we do not know whether BBB is true when we calculate AAA\,
freeing of temporaries after the AAA statement is possible only if we
know at compile time that the context is void\.  Since this is possible
to know only after BBB \(and what follows\) is compiled\, this should be
done at the peep\(\) time\.\.\.

\[Actually\, temporaries \*are\* released anyway at the start of BBB; the
bug is that temporaries generated by AAA are not included at this
step; so the actual fix \(modification of "tide level" of temporary
release inside the BBB block\) will have 0 runtime cost\.\]

Hope this helps\,
Ilya

**Comment by p5pRT** [comment]
### From nospam-abuse@ilyaz.org

On Wed\, Mar 31\, 2004 at 08&#8203;:18&#8203;:21PM \-0800\, Yitzchak Scott\-Thoennes wrote&#8203;:
> On Wed\, Mar 31\, 2004 at 04&#8203;:51&#8203;:44PM \-0800\, Ilya Zakharevich \<nospam\-abuse@&#8203;ilyaz\.org> wrote&#8203;:
> > On Wed\, Mar 31\, 2004 at 11&#8203;:25&#8203;:38AM \-0800\, Yitzchak Scott\-Thoennes wrote&#8203;:
> > > do \{AAA\} and do \{BBB\} was presumably suggested to free temps from AAA
> > > before executing BBB\, but doesn't actually work for that\.
> > 
> > Right\.  It should have been something like&#8203;:
> > 
> >   do \{my $res = \!\!AAA; $res\} and do \{BBB\}
> > 
> > \(sigh; in C it is much easier than in Perl\.\.\.\)\.
> 
> That only works if the overall statement is in void context or AAA is true\.
> Otherwise\, AAA is the resulting value of the statement\.

Oups\, I misread it in my first reply \- as "if BBB is true"\.  First of
all\, the case when AAA is false is much less important; it is not a
very frequent situation when timely destruction of a FALSE value is
important \(has side effects\)\.

But anyway\, your argument is irrelevant; if BBB is entered\,
temporaries created by AAA may \(thus should\) be destroyed\.  If BBB is
not entered\, then there is no problem we discuss at all \- the
temporaries of AAA are timely destroyed even now\.

Yours\,
Ilya

P\.S\. Note that it \*is\* possible to attach DESTROY to a false value;
however\, I can't imagine the usefulness of this\.  \[Is \(0 mod X\)\, where
X is \(a very large?\) modulus\, false?\]

**Comment by p5pRT** [comment]
### From @iabyn


I would like to summarize and restate the two issues\.

A\. freeing tmps

In while\(\) loops etc\, tmps in the conditional are freed before the block
is entered; with if\(\) they are freed at the end of the if statement&#8203;:

&nbsp;    while \(bless \[\]\) \{ \<freed here> \.\.\. \}
&nbsp;    if    \(bless \[\]\) \{ \.\.\. \}  \<freed here>

Q1\. Does anyone disagree that the if\(\) should behave like the while\(\)?
Q2\. Is this worth fixing if it involves a small performance penalty
\(eg an extra flag to the and op or adding an extra nextstate op\)?

B\. Scope of lexicals declared in an if condition

In both if\(my $x \.\.\.\) and while\(my $x \.\.\.\)\, the lexical is visible to the
end of the if/while statement as regards compilation\. At runtime however\,
with the while $x is freed at the end of the statement; with the if it is
freed at the end of the enclosing block&#8203;:

&nbsp;        while \(my $x\.\.\.\) \{ \.\.\.\.\. \} \<$x not visible> \<$x expires>
&nbsp;    \{   if    \(my $x\.\.\.\) \{ \.\.\.\.\. \} \<$x not visible> \.\.\.\.\.\. \}  \<$x expires>

Q3\. Does anyone disagree that the if should behave like the while?
Q4\. Is this worth fixing if it involves a moderate performance penalty
\(eg wrapping the if statement in an enter/leave if there are any lexicals
in its conditional\)?

Dave\.

\-\- 
In my day\, we used to edit the inodes by hand\.  With magnets\.

**Comment by p5pRT** [comment]
### From @hvds

Dave Mitchell \<davem@&#8203;fdisolutions\.com> wrote&#8203;:
:
:I would like to summarize and restate the two issues\.
:
:A\. freeing tmps
:
:In while\(\) loops etc\, tmps in the conditional are freed before the block
:is entered; with if\(\) they are freed at the end of the if statement&#8203;:
:
:    while \(bless \[\]\) \{ \<freed here> \.\.\. \}
:    if    \(bless \[\]\) \{ \.\.\. \}  \<freed here>

At the risk of bearding Occam\, can we add C\<for LIST>\, C\<for \(;;\)> and
optional C\<continue> blocks into this description?

:Q1\. Does anyone disagree that the if\(\) should behave like the while\(\)?

I don't at first glance\, but resulting failures in any existing tests
might change my mind\.

:Q2\. Is this worth fixing if it involves a small performance penalty
:\(eg an extra flag to the and op or adding an extra nextstate op\)?

Given the hotness of pp\_and\, it may be preferable to add a new op rather
than introduce tests into that very short codepath\.

:B\. Scope of lexicals declared in an if condition
:
:In both if\(my $x \.\.\.\) and while\(my $x \.\.\.\)\, the lexical is visible to the
:end of the if/while statement as regards compilation\. At runtime however\,
:with the while $x is freed at the end of the statement; with the if it is
:freed at the end of the enclosing block&#8203;:
:
:        while \(my $x\.\.\.\) \{ \.\.\.\.\. \} \<$x not visible> \<$x expires>
:    \{   if    \(my $x\.\.\.\) \{ \.\.\.\.\. \} \<$x not visible> \.\.\.\.\.\. \}  \<$x expires>
:
:Q3\. Does anyone disagree that the if should behave like the while?
:Q4\. Is this worth fixing if it involves a moderate performance penalty
:\(eg wrapping the if statement in an enter/leave if there are any lexicals
:in its conditional\)?

Roughly the same comments apply\.

Hugo

**Comment by p5pRT** [comment]
### From @iabyn

On Fri\, Apr 02\, 2004 at 02&#8203;:35&#8203;:49PM \+0100\, hv@&#8203;crypt\.org wrote&#8203;:
> Dave Mitchell \<davem@&#8203;fdisolutions\.com> wrote&#8203;:
> :
> :I would like to summarize and restate the two issues\.
> :
> :A\. freeing tmps
> :
> :In while\(\) loops etc\, tmps in the conditional are freed before the block
> :is entered; with if\(\) they are freed at the end of the if statement&#8203;:
> :
> :    while \(bless \[\]\) \{ \<freed here> \.\.\. \}
> :    if    \(bless \[\]\) \{ \.\.\. \}  \<freed here>
> 
> At the risk of bearding Occam\, can we add C\<for LIST>\, C\<for \(;;\)> and
> optional C\<continue> blocks into this description?

Can of worms meet Hugo\, Hugo meet can of worms\.\.\.

A\,B\,C represent tmps&#8203;:

&nbsp;    if    \(B\) \{ \.\.\. \}  \<B freed>
&nbsp;    while \(B\) \{ \<B freed> \.\.\. \}
&nbsp;    while \(B\) \{ \.\.\. \} continue \{ \.\.\.\.; \<B freed> \}

&nbsp;    for \(A\, B\, C\) \{
&nbsp;    \.\.\.
&nbsp;    \}
&nbsp;    \< C\, B\, A freed here>

&nbsp;    for \(A; B; C\) \{
&nbsp;    \<A freed here before first entry to loop>
&nbsp;    \.\.\.\.
&nbsp;    \<C freed>
&nbsp;    \<B freed>
&nbsp;    \}


lexicals&#8203;:

&nbsp;    \{
&nbsp;    if \(my $B\) \{ \.\.\.\.\. \}
&nbsp;    \}
&nbsp;    \<$B expires>


&nbsp;    \{
&nbsp;    while \(my $B\) \{ \.\.\.\.\.  \<$B expires> \}
&nbsp;    \}

&nbsp;    \{
&nbsp;    while \(my $B\) \{ \.\.\.\.\. \}
&nbsp;    continue \{ \.\.\.\.\.\.; \<$C expires> \<$B expires> \}
&nbsp;    \}

&nbsp;    \{
&nbsp;    for \(my $A; my $B; my $C\) \{
&nbsp;        \.\.\.\.

&nbsp;        \<$C expires> \- except on last loop when there is no continue
&nbsp;        \<$B expires>
&nbsp;    \}
&nbsp;    \}
&nbsp;    \<$A expires>

&nbsp;    
\-\- 
The perl5 internals are a complete mess\. It's like Jenga \- to get the
perl5 tower taller and do something new you select a block somewhere in
the middle\, with trepidation pull it out slowly\, and then carefully
balance it somewhere new\, hoping the whole edifice won't collapse as a
result\.
&nbsp;        \- Nicholas Clark\, based on an original by Simon Cozens\.

**Comment by p5pRT** [comment]
### From @ysth

On Fri\, Apr 02\, 2004 at 02&#8203;:21&#8203;:47PM \+0100\, Dave Mitchell \<davem@&#8203;fdisolutions\.com> wrote&#8203;:
> 
> I would like to summarize and restate the two issues\.
> 
> A\. freeing tmps
> 
> In while\(\) loops etc\, tmps in the conditional are freed before the block
> is entered; with if\(\) they are freed at the end of the if statement&#8203;:
> 
>     while \(bless \[\]\) \{ \<freed here> \.\.\. \}
>     if    \(bless \[\]\) \{ \.\.\. \}  \<freed here>
> 
> Q1\. Does anyone disagree that the if\(\) should behave like the while\(\)?
> Q2\. Is this worth fixing if it involves a small performance penalty
> \(eg an extra flag to the and op or adding an extra nextstate op\)?

"extra flag"?  Wouldn't this just involve putting a freetmps in the
else clause in pp\_and?  \(Or per Hugo's suggestion\, the pp\_if that's
otherwise identical to pp\_and\.\)  I assume all the above would apply to
or/unless/pp\_or/new pp\_unless also\.

> B\. Scope of lexicals declared in an if condition
> 
> In both if\(my $x \.\.\.\) and while\(my $x \.\.\.\)\, the lexical is visible to the
> end of the if/while statement as regards compilation\. At runtime however\,
> with the while $x is freed at the end of the statement; with the if it is
> freed at the end of the enclosing block&#8203;:
> 
>         while \(my $x\.\.\.\) \{ \.\.\.\.\. \} \<$x not visible> \<$x expires>
>     \{   if    \(my $x\.\.\.\) \{ \.\.\.\.\. \} \<$x not visible> \.\.\.\.\.\. \}  \<$x expires>
> 
> Q3\. Does anyone disagree that the if should behave like the while?
> Q4\. Is this worth fixing if it involves a moderate performance penalty
> \(eg wrapping the if statement in an enter/leave if there are any lexicals
> in its conditional\)?

**Comment by p5pRT** [comment]
### From @ysth

On Fri\, Apr 02\, 2004 at 01&#8203;:32&#8203;:27AM \-0800\, Ilya Zakharevich \<nospam\-abuse@&#8203;ilyaz\.org> wrote&#8203;:
> On Wed\, Mar 31\, 2004 at 08&#8203;:18&#8203;:21PM \-0800\, Yitzchak Scott\-Thoennes wrote&#8203;:
> > On Wed\, Mar 31\, 2004 at 04&#8203;:51&#8203;:44PM \-0800\, Ilya Zakharevich \<nospam\-abuse@&#8203;ilyaz\.org> wrote&#8203;:
> > > On Wed\, Mar 31\, 2004 at 11&#8203;:25&#8203;:38AM \-0800\, Yitzchak Scott\-Thoennes wrote&#8203;:
> > > > do \{AAA\} and do \{BBB\} was presumably suggested to free temps from AAA
> > > > before executing BBB\, but doesn't actually work for that\.
> > > 
> > > Right\.  It should have been something like&#8203;:
> > > 
> > >   do \{my $res = \!\!AAA; $res\} and do \{BBB\}
> > > 
> > > \(sigh; in C it is much easier than in Perl\.\.\.\)\.
> > 
> > That only works if the overall statement is in void context or AAA is true\.
> > Otherwise\, AAA is the resulting value of the statement\.
> 
> Oups\, I misread it in my first reply \- as "if BBB is true"\.  First of
> all\, the case when AAA is false is much less important; it is not a
> very frequent situation when timely destruction of a FALSE value is
> important \(has side effects\)\.

"AAA is true" was meant as "BBB is entered"\.  I was trying to list the
cases where your \!\!AAA will produce the same result\.

**Comment by p5pRT** [comment]
### From @chipdude

According to Yitzchak Scott\-Thoennes&#8203;:
> Wouldn't this just involve putting a freetmps in the else clause in
> pp\_and?

But what if the left operand of the pp\_and is itself a temp?
\-\- 
Chip Salzenberg               \- a\.k\.a\. \-               \<chip@&#8203;pobox\.com>
"I wanted to play hopscotch with the impenetrable mystery of existence\,
&nbsp;    but he stepped in a wormhole and had to go in early\."  // MST3K

**Comment by p5pRT** [comment]
### From @chipdude

According to Yitzchak Scott\-Thoennes&#8203;:
> Wouldn't this just involve putting a freetmps in the else clause in
> pp\_and?

\[Please ignore my last message\, I missed the "else clause" phrase\.\]

Consider this&#8203;:

&nbsp;    SelectSaver\->new\(\\\*HANDLE\)
&nbsp;      and print $this\_goes\_to\_handle;

Do we want to break this?  Does it matter?
\-\- 
Chip Salzenberg               \- a\.k\.a\. \-               \<chip@&#8203;pobox\.com>
"I wanted to play hopscotch with the impenetrable mystery of existence\,
&nbsp;    but he stepped in a wormhole and had to go in early\."  // MST3K

**Comment by p5pRT** [comment]
### From @ysth

On Thu\, Apr 01\, 2004 at 12&#8203;:03&#8203;:17PM \-0800\, Ilya Zakharevich \<nospam\-abuse@&#8203;ilyaz\.org> wrote&#8203;:
> On Wed\, Mar 31\, 2004 at 08&#8203;:18&#8203;:21PM \-0800\, Yitzchak Scott\-Thoennes wrote&#8203;:
> > On Wed\, Mar 31\, 2004 at 04&#8203;:51&#8203;:44PM \-0800\, Ilya Zakharevich \<nospam\-abuse@&#8203;ilyaz\.org> wrote&#8203;:
> > > On Wed\, Mar 31\, 2004 at 11&#8203;:25&#8203;:38AM \-0800\, Yitzchak Scott\-Thoennes wrote&#8203;:
> > > > do \{AAA\} and do \{BBB\} was presumably suggested to free temps from AAA
> > > > before executing BBB\, but doesn't actually work for that\.
> > > 
> > > Right\.  It should have been something like&#8203;:
> > > 
> > >   do \{my $res = \!\!AAA; $res\} and do \{BBB\}
> > > 
> > > \(sigh; in C it is much easier than in Perl\.\.\.\)\.
> 
> > That only works if the overall statement is in void context or AAA is true\.
> > Otherwise\, AAA is the resulting value of the statement\.
> 
> Do you still consider
> 
>   if \(AAA\) \{BBB\}
> 
> a statement?  It is not\.  AAA is a separate statement; BBB consists of
> separate statements\.

I meant from a semantic point of view\, i\.e\. my "statement" is perly\.y's
"line"\, distinct from "expr"\.  I'm not sure in what context you call
"if \(AAA\) \{BBB\}" not a statement\, but it probably doesn't matter\.

> But anyway\, your analysis finally pointed out \*some\* reason why the
> current behaviour makes sense in \*some\* situations\.  Obviously\, I
> missed this; thanks\.
> 
> Since we do not know whether BBB is true when we calculate AAA\,
> freeing of temporaries after the AAA statement is possible only if we
> know at compile time that the context is void\.  Since this is possible
> to know only after BBB \(and what follows\) is compiled\, this should be
> done at the peep\(\) time\.\.\.
> 
> \[Actually\, temporaries \*are\* released anyway at the start of BBB; the
> bug is that temporaries generated by AAA are not included at this
> step; so the actual fix \(modification of "tide level" of temporary
> release inside the BBB block\) will have 0 runtime cost\.\]

I'm afraid I didn't follow the rest of this\.  What change would you
make at compile time?  Can you show pseudocode?

**Comment by p5pRT** [comment]
### From @iabyn

On Fri\, Apr 02\, 2004 at 11&#8203;:27&#8203;:41AM \-0800\, Yitzchak Scott\-Thoennes wrote&#8203;:
> On Fri\, Apr 02\, 2004 at 02&#8203;:21&#8203;:47PM \+0100\, Dave Mitchell \<davem@&#8203;fdisolutions\.com> wrote&#8203;:
> > 
> > I would like to summarize and restate the two issues\.
> > 
> > A\. freeing tmps
> > 
> > In while\(\) loops etc\, tmps in the conditional are freed before the block
> > is entered; with if\(\) they are freed at the end of the if statement&#8203;:
> > 
> >     while \(bless \[\]\) \{ \<freed here> \.\.\. \}
> >     if    \(bless \[\]\) \{ \.\.\. \}  \<freed here>
> > 
> > Q1\. Does anyone disagree that the if\(\) should behave like the while\(\)?
> > Q2\. Is this worth fixing if it involves a small performance penalty
> > \(eg an extra flag to the and op or adding an extra nextstate op\)?
> 
> "extra flag"?  Wouldn't this just involve putting a freetmps in the
> else clause in pp\_and?  \(Or per Hugo's suggestion\, the pp\_if that's
> otherwise identical to pp\_and\.\)  I assume all the above would apply to
> or/unless/pp\_or/new pp\_unless also\.

This was just vague handwaving\. I was thinking of a flag in OP\_AND that
distingushes between if\(A\)\{B\} and A&&B \(if they need distinguishing\)\.
Or as Hugo points out\, have a separate op if necessary\.
I haven't thought throught the implementation details at all; I was
seeing if we could get a concensus on the desired behaviour first\.

\-\- 
The warp engines start playing up a bit\, but seem to sort themselves out
after a while without any intervention from boy genius Wesley Crusher\.
&nbsp;    \-\- Things That Never Happen in "Star Trek" \#17

**Comment by p5pRT** [comment]
### From @hvds

Chip Salzenberg \<chip@&#8203;pobox\.com> wrote&#8203;:
:Consider this&#8203;:
:
:    SelectSaver\->new\(\\\*HANDLE\)
:      and print $this\_goes\_to\_handle;
:
:Do we want to break this?  Does it matter?

Hmm\, good point\. I use a similar approach frequently \(eg to switch
databases\, or change umask\)\, but it never occurred to me to use
anything less than a full block scope to protect it&#8203;:
&nbsp;  \{
&nbsp;    my $saver = set\_umask\(0\);
&nbsp;    mkdir\($public\, 0777\) or die "mkdir $public&#8203;: $\!\\n";
&nbsp;  \}

If people are relying on freetmps timing for this sort of thing\,
I'm tempted to say they lose unless they can point to perl core
documentation promising the timing they want\. But it's an insidious
sort of bug to be springing on them \(or at least\, an insidious bug
in their code to be exposing\)\.

Hugo

**Comment by p5pRT** [comment]
### From nospam-abuse@ilyaz.org

On Fri\, Apr 02\, 2004 at 11&#8203;:39&#8203;:15AM \-0800\, Yitzchak Scott\-Thoennes wrote&#8203;:
> > Do you still consider
> > 
> >   if \(AAA\) \{BBB\}
> > 
> > a statement?  It is not\.  AAA is a separate statement; BBB consists of
> > separate statements\.
> 
> I meant from a semantic point of view\, i\.e\. my "statement" is perly\.y's
> "line"\, distinct from "expr"\.  I'm not sure in what context you call
> "if \(AAA\) \{BBB\}" not a statement\, but it probably doesn't matter\.

So you meant it is not in semantic\, but in the implementation
\(perly\.y\) point of view\.  I meant it in the \*semantic\* ;\-\) point of
view&#8203;: "you should not expect that temporaries are released before the
beginning of the following statement"\.

> > But anyway\, your analysis finally pointed out \*some\* reason why the
> > current behaviour makes sense in \*some\* situations\.  Obviously\, I
> > missed this; thanks\.
> > 
> > Since we do not know whether BBB is true when we calculate AAA\,
> > freeing of temporaries after the AAA statement is possible only if we
> > know at compile time that the context is void\.  Since this is possible
> > to know only after BBB \(and what follows\) is compiled\, this should be
> > done at the peep\(\) time\.\.\.
> > 
> > \[Actually\, temporaries \*are\* released anyway at the start of BBB; the
> > bug is that temporaries generated by AAA are not included at this
> > step; so the actual fix \(modification of "tide level" of temporary
> > release inside the BBB block\) will have 0 runtime cost\.\]
> 
> I'm afraid I didn't follow the rest of this\.  What change would you
> make at compile time?  Can you show pseudocode?

I did not know what I was talking about; I \(very uncritically ;\-\) took
what you said on faith\, while misinterpreting what you said\.  Very sad
situation from a semantic point of view\.

Here is my current reading of the situation with the current design of
temporaries release&#8203;:

&nbsp;  in any sequence of statements

&nbsp;   \{    \# point EnTeR\_AAA
&nbsp;     AAA1; AAA2; AAA3;
&nbsp;   \}

the temporaries are allowed to accumulate \(on a special stack\) during
each of AAAn\.  However\, perl notices the amount of temporaries when it
reaches "point EnTeR"\, and frees the temporaries \*after this point\* at
the beginning of each AAAn \(this is the "tide mark" in one of my
mails\)\.  So the thing compiles to

&nbsp;   \{    \# memorize point EnTeR\_AAA on the tmps\-stack
&nbsp;     Free tmps after EnTeR\_AAA
&nbsp;     AAA1;
&nbsp;     Free tmps after EnTeR\_AAA
&nbsp;     AAA2;
&nbsp;     Free tmps after EnTeR\_AAA
&nbsp;     AAA3;
&nbsp;   \}

With the current implementation of

&nbsp;  \{
&nbsp;   AAA1;
&nbsp;   if \(AAA2\) \{BBB1; BBB2; BBB3\}
&nbsp;   AAA3;
&nbsp;  \}

things \(apparently\) look like this

&nbsp;   \{    \# memorize point EnTeR\_AAA on the tmps\-stack
&nbsp;     Free tmps after EnTeR\_AAA
&nbsp;     AAA1;
&nbsp;     Free tmps after EnTeR\_AAA
&nbsp;     AAA2 or goto end\_AAA2;
&nbsp;     \{    \# memorize point EnTeR\_BBB on the tmps\-stack
&nbsp;       Free tmps after EnTeR\_BBB
&nbsp;       BBB1;
&nbsp;       Free tmps after EnTeR\_BBB
&nbsp;       BBB2;
&nbsp;       Free tmps after EnTeR\_BBB
&nbsp;       BBB3;
&nbsp;     \}
&nbsp;   end\_AAA2&#8203;:
&nbsp;     Free tmps after EnTeR\_AAA
&nbsp;     AAA3;
&nbsp;   \}

All I say now that one can replace the point EnTeR\_BBB by the point
EnTeR\_AAA without any harm; however\, this will free the temporaries
introduces by AAA2 at the beginning of BBB1\.

Hope this helps\,
Ilya

P\.S\.  Actually\, I do not know whether the temporaries are actually
free\(\)ed at the beginning of AAA1 and BBB1 \(but they definitely are at
the beginning of AAA2/3 and BBB2/3\)\.  If they are NOT\, then this
addresses the example of Chip's \- it continues to work as it did
without distinguishing if \(AAA\) \{BBB\} and \`AAA and BBB'\.

**Comment by p5pRT** [comment]
### From @iabyn

On Sat\, Apr 03\, 2004 at 01&#8203;:21&#8203;:20AM \+0100\, hv@&#8203;crypt\.org wrote&#8203;:
> Chip Salzenberg \<chip@&#8203;pobox\.com> wrote&#8203;:
> :Consider this&#8203;:
> :
> :    SelectSaver\->new\(\\\*HANDLE\)
> :      and print $this\_goes\_to\_handle;
> :
> :Do we want to break this?  Does it matter?
> 
> Hmm\, good point\. I use a similar approach frequently \(eg to switch
> databases\, or change umask\)\, but it never occurred to me to use
> anything less than a full block scope to protect it&#8203;:
>   \{
>     my $saver = set\_umask\(0\);
>     mkdir\($public\, 0777\) or die "mkdir $public&#8203;: $\!\\n";
>   \}
> 
> If people are relying on freetmps timing for this sort of thing\,
> I'm tempted to say they lose unless they can point to perl core
> documentation promising the timing they want\. But it's an insidious
> sort of bug to be springing on them \(or at least\, an insidious bug
> in their code to be exposing\)\.

I'd like to draw a clear distinction between C\<if\(A\)\{B\}> and \<A and B>\.
The former is a single expression and tmps shouldn't be freed until the end
of it\. I only expect the tmps to be freed early in the 'if' case\.
\-\- 
You live and learn \(although usually you just live\)\.

**Comment by p5pRT** [comment]
### From @iabyn

On Sat\, Apr 03\, 2004 at 12&#8203;:06&#8203;:36PM \+0100\, Dave Mitchell wrote&#8203;:
> I'd like to draw a clear distinction between C\<if\(A\)\{B\}> and \<A and B>\.
> The former is a single expression and tmps shouldn't be freed until the end
> of it\. I only expect the tmps to be freed early in the 'if' case\.

s/former/latter/ \!

\-\- 
"There's something wrong with our bloody ships today\, Chatfield\."
&nbsp;    \-\- Admiral Beatty at the Battle of Jutland\, 31st May 1916\.

**Comment by p5pRT** [comment]
### From @ysth

On Fri\, Apr 02\, 2004 at 11&#8203;:53&#8203;:15PM \-0800\, Ilya Zakharevich \<nospam\-abuse@&#8203;ilyaz\.org> wrote&#8203;:
> With the current implementation of
> 
>   \{
>    AAA1;
>    if \(AAA2\) \{BBB1; BBB2; BBB3\}
>    AAA3;
>   \}
> 
> things \(apparently\) look like this
> 
>    \{    \# memorize point EnTeR\_AAA on the tmps\-stack
>      Free tmps after EnTeR\_AAA
>      AAA1;
>      Free tmps after EnTeR\_AAA
>      AAA2 or goto end\_AAA2;
>      \{    \# memorize point EnTeR\_BBB on the tmps\-stack
>        Free tmps after EnTeR\_BBB
>        BBB1;
>        Free tmps after EnTeR\_BBB
>        BBB2;
>        Free tmps after EnTeR\_BBB
>        BBB3;
>      \}
>    end\_AAA2&#8203;:
>      Free tmps after EnTeR\_AAA
>      AAA3;
>    \}
> 
> All I say now that one can replace the point EnTeR\_BBB by the point
> EnTeR\_AAA without any harm; however\, this will free the temporaries
> introduces by AAA2 at the beginning of BBB1\.
> 
> Hope this helps\,
> Ilya
> 
> P\.S\.  Actually\, I do not know whether the temporaries are actually
> free\(\)ed at the beginning of AAA1 and BBB1 \(but they definitely are at
> the beginning of AAA2/3 and BBB2/3\)\.  If they are NOT\, then this
> addresses the example of Chip's \- it continues to work as it did
> without distinguishing if \(AAA\) \{BBB\} and \`AAA and BBB'\.

Not sure how that addresses Chip's example \(if you mean the SelectSaver thing\)\.

It appears to a quick glance as if the nextstate is optimized away in a scope
block but not an enter/leave block\, which seems odd\.

**Comment by p5pRT** [comment]
### From @ysth

On Sun\, Apr 04\, 2004 at 01&#8203;:41&#8203;:34AM \-0800\, Yitzchak Scott\-Thoennes \<sthoenna@&#8203;efn\.org> wrote&#8203;:
> On Fri\, Apr 02\, 2004 at 11&#8203;:53&#8203;:15PM \-0800\, Ilya Zakharevich \<nospam\-abuse@&#8203;ilyaz\.org> wrote&#8203;:
> > All I say now that one can replace the point EnTeR\_BBB by the point
> > EnTeR\_AAA without any harm; however\, this will free the temporaries
> > introduces by AAA2 at the beginning of BBB1\.

Forgot to say&#8203;: that's quite clever; not sure how to actually accomplish
it while having e\.g\. local within the if's block restored at the right
point\.

Thanks for the detailed demonstration\.
